execution_engine/src/shared/host_function_costs.rs

```diff
@@ -130,14 +131,17 @@ where
     }
 
     /// Calculate gas cost for a host function
-    pub fn calculate_gas_cost(&self, weights: T) -> Gas {
+    pub fn calculate_gas_cost(&self, weights: T) -> Option<Gas> {
         let mut gas = Gas::new(self.cost.into());
+
+        // sum i->n weight_i * arg_i
         for (argument, weight) in self.arguments.as_ref().iter().zip(weights.as_ref()) {
             let lhs = Gas::new((*argument).into());
             let rhs = Gas::new((*weight).into());
-            gas += lhs * rhs;
+            let pn = lhs.checked_mul(&rhs)?;
```

Comment:
Rather than an return an option, shouldn't we error with `Err(Error::GasLimit)` and return a `Result` instead?